benchmark/Streamly/Benchmark/FileSystem/DirIO.hs

```diff
@@ -0,0 +1,153 @@
+    . StreamK.fromStream
+
+streamDir :: Either Path b -> Stream IO (Either Path Path)
+streamDir = either Dir.readEitherPaths (const Stream.nil)
+
+unfoldDir :: Unfold IO (Either Path b) (Either Path Path)
+unfoldDir = Unfold.either Dir.eitherReaderPaths Unfold.nil
+
+_streamDirMaybe :: Either Path b -> Maybe (Stream IO (Either Path Path))
+_streamDirMaybe = either (Just . Dir.readEitherPaths) (const Nothing)
+
+_streamDirByteChunked
+    :: Either [Path] b -> Stream IO (Either [Path] (Array Word8))
+_streamDirByteChunked = either Dir.readEitherByteChunks (const Stream.nil)
+
+_streamDirByteChunkedMaybe
+    :: Either [Path] b -> Maybe (Stream IO (Either [Path] (Array Word8)))
+_streamDirByteChunkedMaybe =
+    either (Just . Dir.readEitherByteChunks) (const Nothing)
+
+_streamDirChunked :: Either [Path] b -> Stream IO (Either [Path] [Path])
+_streamDirChunked = either Dir.readEitherChunks (const Stream.nil)
+
+streamDirChunkedMaybe :: Either [Path] b -> Maybe (Stream IO (Either [Path] [Path]))
+streamDirChunkedMaybe = either (Just . Dir.readEitherChunks) (const Nothing)
+
+#if !defined(mingw32_HOST_OS) && !defined(__MINGW32__)
+-- Fastest implementation, only works for posix as of now.
+listDirByteChunked :: FilePath -> IO ()
+listDirByteChunked inp = do
+    Stream.fold Fold.drain
+        $ Stream.catRights
+        $ Stream.concatIterateDfs streamDirChunkedMaybe
+        $ Stream.fromPure (Left [fromJust $ Path.fromString inp])
+#endif
+
+-- Faster than the listDir implementation below
+listDirChunked :: FilePath -> IO ()
+listDirChunked inp = do
+    Stream.fold Fold.drain
+        $ Stream.unfoldEachEndBy 10 Array.reader
+        $ fmap Path.toChunk
+        $ Stream.unfoldEach Unfold.fromList
+        $ fmap (either id id)
+        $ Stream.concatIterateDfs streamDirChunkedMaybe
+        $ Stream.fromPure (Left [fromJust $ Path.fromString inp])
+
+listDir :: FilePath -> IO ()
+listDir inp = do
+    Stream.fold Fold.drain
+        $ Stream.unfoldEachEndBy 10 Array.reader
+        $ fmap (Path.toChunk . either id id)
+        $ Stream.unfoldIterateDfs unfoldDir
+        $ Stream.fromPure (Left (fromJust $ Path.fromString inp))
+
+--------------------------------------------------------------------------------
+-- Main
+--------------------------------------------------------------------------------
+
+moduleName :: String
+moduleName = "FileSystem.DirIO"
+
+createDirStucture :: IO FilePath
+createDirStucture = do
+    let dirRoot = scratchDir ++ "dir-structure"
+    let cmd =
+            "benchmark/create_dir_structure.sh  " ++ dirRoot ++ " 5 5"
+    callCommand ("mkdir -p " ++ dirRoot)
+    callCommand cmd
+    return dirRoot
+
+-- | List the current directory recursively
+main :: IO ()
+main = do
+    bigDirStructure <- createDirStucture
+    defaultMain
+        [ bgroup (o_1_space_prefix moduleName)
+            [ bench "listDirChunked" $ nfIO $ listDirChunked bigDirStructure
```

Comment:
mention in the benchmark description which APIs are being tested e.g. `concateIterateDfs . readEitherPaths` etc.